doc/reference/trd-time.md

```diff
@@ -28,148 +29,342 @@ broad categories: alarms and timers. Alarms continuously increment a clock and
+  fn set_overflow_client(&'a self, &'a dyn OverflowClient);
 }
 ```
 
-The `disable` method disables the time controller, stopping any pending alarms
-or timers. It MAY disable the underlying clock to save power.
+The `OverflowClient` trait is separated from the `AlarmClient` trait
+because there are cases when software simply wants a free-running
+counter to keep track of time, but does not need triggers at a
+particular time. For hardware that has a limited number of
+compare registers, allocating one of them when the compare itself
+isn't needed would be wasteful.
 
-The `is_armed` methods indicates whether the time controller has an event
-queued (i.e. an alarm or timer).
+A `Counter` implementation MUST NOT provide a `Frequency` of a higher
+resolution than an underlying hardware counter. For example, if the
+underlying hardware counter has a frequency of 32kHz, then a `Counter`
+cannot say it has a frequency of 1MHz by multiplying the underlying
+counter by 32. A `Counter` implementation MAY provide a `Frequency` of
+a lower resolution (e.g., by stripping bits).
 
-[associated type]: https://doc.rust-lang.org/book/associated-types.html
+The `reset` method of `Counter` resets the counter to 0.
 
-3 `Alarm` trait
+4 `Alarm` and `AlarmClient` traits
 ===============================
 
-Instances of the `Alarm` trait track a continuously incrementing clock and can
-fire an event when the clock reaches a specific value. The trait is a subtype
-of the `Time` trait and, as a result, has access to the `Time::Frequency`
-associated type. Whenever the trait referes to "tics", they are interpreted in
-native clock tics.
+Instances of the `Alarm` trait track an incrementing clock and can
+trigger callbacks when the clock reaches a specific value as well as
+when it overflows. The trait is derived from `Time` trait and
+therefore has associated `Time::Frequency` and `Ticks` types.
 
-```rust
-pub trait Alarm: Time {
-    fn now(&self) -> u32;
+The `AlarmClient` trait handles callbacks from an instance of `Alarm`.
+The trait derives from `OverflowClient` and adds an additional callback
+denoting that the time specified to the `Alarm` has been reached.
 
-    fn set_alarm(&self, tics: u32);
+`Alarm` and `Timer` (presented below) differ in their level of
+abstraction. An `Alarm` presents the abstraction of receiving a
+callback when a point in time is reached or on an overflow. In
+contrast, `Timer` allows one to request callbacks at some interval in
+the future, either once or periodically. `Alarm` requests a callback
+at an absolute moment while `Timer` requests a callback at a point
+relative to now.
 
-    fn get_alarm(&self) -> u32;
+```rust
+pub trait AlarmClient: OverflowClient {
+  fn alarm(&self);
+}
+
+pub trait Alarm: Time {
+  fn set_alarm(&self, reference: Self::Ticks, dt: Self::Ticks);
+  fn get_alarm(&self) -> Self::Ticks;
+  fn disarm(&self) -> ReturnCode;
+  fn set_alarm_client(&'a self, client: &'a dyn AlarmClient);
```

Comment:
Why not just set_client() like all the others?

Reply:
Good question. It's because of cases in which you want to handle both overflows (from Counter) and compares (from Alarm). E.g, if a structure  If they are named the same thing it's a pain. Basically, for these abstractions it is quite possible that a single structure implements multiple traits, and so easily disambiguating their methods makes things easier.

Reply:
Is it not worth naming the other ones this way as well? It seems strange that this set_client has a different name to all the others in Tock.

Reply:
It really has to do with the fact that this set of traits are tightly entwined and it's quite normal for a single struct to implement multiple of them. I don't recall ever seeing this issue before in Tock, although it might have happened and so I missed it. If we wanted to change all of the set_clients I wouldn't be against it, but I don't think it's necessary. The analogy for me in natural language is when you have two things you disambiguate them with an adjective, but when you have one thing you don't ("the smaller binary" vs "the binary").